From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltarsCN <337293323+Altars3668@users.noreply.github.com>
Date: Mon, 31 Aug 2026 13:32:42 -0400
Subject: [PATCH] configure: Regenerate for the new srpapi makefile.

(cherry picked from commit 361d53a50d120f228bee5e6e25eccb823c6be623)

diff --git a/configure b/configure
index e73491f3289..0d553eee4e5 100755
--- a/configure
+++ b/configure
@@ -1463,6 +1463,7 @@ enable_softpub
 enable_spoolss
 enable_sppc
 enable_srclient
+enable_srpapi
 enable_srvcli
 enable_srvsvc
 enable_sspicli
@@ -22747,6 +22748,7 @@ wine_fn_config_makefile dlls/spoolss enable_spoolss
 wine_fn_config_makefile dlls/spoolss/tests enable_tests
 wine_fn_config_makefile dlls/sppc enable_sppc
 wine_fn_config_makefile dlls/srclient enable_srclient
+wine_fn_config_makefile dlls/srpapi enable_srpapi
 wine_fn_config_makefile dlls/srvcli enable_srvcli
 wine_fn_config_makefile dlls/srvsvc enable_srvsvc
 wine_fn_config_makefile dlls/sspicli enable_sspicli
