--- a/dlls/ntdll/version.c
+++ b/dlls/ntdll/version.c
@@ -56,6 +56,7 @@
     WIN81,   /* Windows 8.1 */
     WIN10,   /* Windows 10 */
     WIN11,   /* Windows 11 */
+    WIN11_24H2, /* Windows 11 24H2 */
     NB_WINDOWS_VERSIONS
 } WINDOWS_VERSION;
 
@@ -176,6 +177,11 @@
         sizeof(RTL_OSVERSIONINFOEXW), 10, 0, 22000, VER_PLATFORM_WIN32_NT,
         L"", 0, 0, VER_SUITE_SINGLEUSERTS, VER_NT_WORKSTATION, 0
     },
+    /* WIN11_24H2 */
+    {
+        sizeof(RTL_OSVERSIONINFOEXW), 10, 0, 26100, VER_PLATFORM_WIN32_NT,
+        L"", 0, 0, VER_SUITE_SINGLEUSERTS, VER_NT_WORKSTATION, 0
+    },
 };
 
 static const struct { WCHAR name[12]; WINDOWS_VERSION ver; } version_names[] =
@@ -207,6 +213,7 @@
     { L"win81", WIN81 },
     { L"win10", WIN10 },
     { L"win11", WIN11 },
+    { L"win11_24h2", WIN11_24H2 },
 };
 
 
